File: TradeItIosEmsApi/TradeItUserAgentProvider.m
#import "TradeItUserAgentProvider.h"
#include <UIKit/UIKit.h>
#include <sys/sysctl.h>

#ifdef CARTHAGE
    #import <TradeItIosTicketSDK2Carthage/TradeItIosTicketSDK2Carthage-Swift.h>
#else
    #import <TradeItIosTicketSDK2/TradeItIosTicketSDK2-Swift.h>
#endif

@implementation TradeItUserAgentProvider

+ (NSString *)getUserAgent {
    static NSString *userAgent = nil;
    static dispatch_once_t onceToken;

    dispatch_once(&onceToken, ^{
        NSDictionary<NSString *, id> *bundleDict = [[NSBundle mainBundle] infoDictionary];
        NSString *appName = [bundleDict valueForKey:@"CFBundleName"];
        NSString *appVersion = [bundleDict valueForKey:@"CFBundleShortVersionString"];

        NSString *appDescriptor = [NSString stringWithFormat:@"%@/%@", appName, appVersion];

        UIDevice *device = [UIDevice currentDevice];
        NSString *systemVersion = [device systemVersion];

        NSString *osDescriptor = [NSString stringWithFormat:@"%@ %@", @"iOS", systemVersion];

        NSString *hardwareString = [self getSysInfoByName:"hw.model"];

        NSDictionary<NSString *, id> *sdkBundleInfoDictionary = [[TradeItBundleProvider provide] infoDictionary];

        NSString *sdkName = [sdkBundleInfoDictionary valueForKey:@"CFBundleName"];
        NSString *sdkVersion = [sdkBundleInfoDictionary valueForKey:@"CFBundleShortVersionString"];

        userAgent = [NSString stringWithFormat:@"%@/%@ (%@) / %@/%@", appDescriptor,  osDescriptor, hardwareString, sdkName, sdkVersion];
    });

    return userAgent;
}

+ (NSString *)getSysInfoByName:(char *)typeSpecifier {
    size_t size;
    sysctlbyname(typeSpecifier, NULL, &size, NULL, 0);
    
    char *answer = malloc(size);
    sysctlbyname(typeSpecifier, answer, &size, NULL, 0);
    
    NSString *results = [NSString stringWithCString:answer encoding: NSUTF8StringEncoding];
    
    free(answer);
    return results;
}

@end

File: TradeItIosEmsApi/TradeItUserAgentProvider.swift
class TradeItUserAgentProvider: NSObject {

    static let userAgent: String = {
        var bundleDict = Bundle.main.infoDictionary
        var appName = bundleDict?["CFBundleName"] as? String
        var appVersion = bundleDict?["CFBundleShortVersionString"] as? String
        
        var appDescriptor = "\(appName ?? "")/\(appVersion ?? "")"
        
        var device = UIDevice.current
        var systemVersion = device.systemVersion
        
        var osDescriptor = "\("iOS") \(systemVersion)"
        
        var hardwareString = getSysInfo(byName: "hw.model")
        
        var sdkBundleInfoDictionary = TradeItBundleProvider.provide().infoDictionary
        
        var sdkName = sdkBundleInfoDictionary?["CFBundleName"] as? String
        var sdkVersion = sdkBundleInfoDictionary?["CFBundleShortVersionString"] as? String
        
        return "\(appDescriptor)/\(osDescriptor) (\(String(describing: hardwareString))) / \(sdkName ?? "")/\(sdkVersion ?? "")"
    }()
    
    private static func getSysInfo(byName typeSpecifier: String?) -> String? {
        var size: size_t = 0
        sysctlbyname(typeSpecifier, nil, &size, nil, 0)
        
        var answer = [CChar](repeating: 0,  count: size)
        sysctlbyname(typeSpecifier, &answer, &size, nil, 0)
        
        let results = String(cString: answer, encoding: .utf8)
        return results
    }
    
}
